### PR TITLE
Regex language injection

### PR DESCRIPTION
**Pull request by saul**
image

**Todo**:

- [x] Write tests
- [ ] Fix regular string support - escape characters are seemingly ignored
- [x] Support RegExpPattern attribute
- [x] Support completion 
- [x] Support Regex Provider
- [x] Support injection by comment
- [x] Support StringSyntax attribute
- [x] Support RegularExpression attribute
- [ ] Support injection into assignments

### CODE REVIEW

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Injected/FSharpLiteralInjectionTarget.fs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                    ClrRegexLexerOptions(true)
+                | _ ->
+                    ClrRegexLexerOptions.None
+
+            CSharpRegExpBuffer(StringBuffer(text), literalType, lexerOptions) :> _
+
+        override _.DoNotProcessNodeInterior _ = false
+
+        override _.IsPrimaryLanguageApplicable(sourceFile) =
+            sourceFile.LanguageType :? FSharpProjectFileType
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/Injected/FSharpLiteralInjectionTarget.fs:86]:
```suggestion
            sourceFile.LanguageType.Is<FSharpProjectFileType>()
```

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Injected/FSharpLiteralInjectionTarget.fs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+        override _.AllowsLineBreaks _ = true
+
+        override _.IsWhitespaceToken(token) =
+            token.GetTokenType().IsWhitespace
+
+        // Used only in JS/JSON for now
+        override x.FixValueRangeForLiteral(element) =
+            let startOffset = (x :> IInjectionTargetLanguage).GetStartOffsetForString(element)
+            element.GetTreeTextRange().TrimLeft(startOffset).TrimRight(1)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi.Features/src/Injected/FSharpLiteralInjectionTarget.fs:99]: Do we need this? We're not going to inject js/json on backend AFAIK, is it needed for something else?

File: ReSharper.FSharp/src/FSharp.Psi/src/Annotations/FSharpAnnotatedCacheProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      private readonly AnnotatedEntitiesSet myContext;
+
+      public Processor(HashSet<string> attributeNames, AnnotatedEntitiesSet context)
+      {
+        myAttributeNames = attributeNames;
+        myContext = context;
+      }
+
+      public bool ProcessingIsFinished => false;
+      public bool InteriorShouldBeProcessed(ITreeNode element) => element is not IChameleonNode;
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Annotations/FSharpAnnotatedCacheProcessor.cs:35]: We should probably also cover cases where parameters are inside a chameleon.
```fsharp
let f x =
    fun ([<A>] y) ->
        ()
```
and
```fsharp
let f =
    fun ([<A>] y) ->
        ()
```

That would need an addition to `ChameleonExpressionUtil` too.

File: ReSharper.FSharp/src/FSharp.Psi/src/Annotations/FSharpAnnotatedCacheProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        var defaultMemberName = typeDeclaration.DeclaredName;
+        VisitParametersOwner(primaryConstructorDeclaration, defaultMemberName);
+
+        if (AttributeUtil.HasAttributeSuffix(defaultMemberName, out var nameWithoutAttributeSuffix))
+          VisitParametersOwner(primaryConstructorDeclaration, nameWithoutAttributeSuffix);
+      }
+
+      public override void VisitTopBinding(ITopBinding topBinding)
+      {
+        foreach (var declaration in topBinding.HeadPattern.Declarations)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/Annotations/FSharpAnnotatedCacheProcessor.cs:74]: `HeadPattern` may be `null`.

**Review by auduchinok — COMMENTED** [review]
@DedSec256 Very nice work!

I've left some comments. Hope it helps :)

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Injected/FSharpLiteralInjectionTarget.fs (hunk 2 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/Injected/FSharpLiteralInjectionTarget.fs:99]: I think we can remove this code

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Injected/FSharpRegexProviders.fs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+type RegExprPsiProvider(injectorProvider: FSharpRegexInjectionProvider) =
+    inherit LiteralsInjectionPsiProvider<FSharpLanguage, ClrRegexLanguage>(injectorProvider, ClrRegexLanguage.Instance)
+
+    override _.ProvidedLanguageCanHaveNestedInjects = false
+
+
+[<Language(typeof<FSharpLanguage>)>]
+type FSharpRegularExpressionCompletionProvider() =
+    interface IRegexLanguageSpecificCompletionProvider with
+        override this.InitializeContext(owner, _) = owner :? ILiteralExpression
```
[the remark this answers is not in the document]
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/Injected/FSharpRegexProviders.fs:43]: This code is invoked for nodes in which injections have already been checked. However, one really needs to add a check for ILiteralPat

[the remark this answers is not in the document]
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/Injected/FSharpRegexProviders.fs:43]: This code is invoked only for nodes checked for injection earlier. This automatically works for all applicable FSharpString, including active patterns and type providers.

[the remark this answers is not in the document]
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Psi.Features/src/Injected/FSharpRegexProviders.fs:43]: Can it then always return `true`? Why does it only check the expressions?

File: ReSharper.FSharp/src/FSharp.Psi/src/Annotations/FSharpAnnotatedCacheProcessor.cs (hunk 3 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp.Psi/src/Annotations/FSharpAnnotatedCacheProcessor.cs:35]: We also need to support probably the same case for ILocalBinding. And a simpler case with the regular local binding function.

**Comment by DedSec256** [comment]
- [ ] FSharpAnnotatedMembersCacheProcessor (signatures/accessors, etc)

**Review by auduchinok — APPROVED** [review]
This is a really good work! Thanks @DedSec256 and @saul!

**Pull request merged**